py-rse/style.Rmd

```diff
@@ -59,165 +48,228 @@ Look for algorithmic problems first.
-You should only use [`CamelCase`][camel-case] for classes,
-which are outside the scope of this lesson.
+**Handle special cases at the start of the function.**
 
-### Put two blank links between each function definition.
+This helps readers mentally get them out of the way
+and focus on the "normal" case.
+For example, this:
 
-This helps them stand out.
+```python
+def positive_sum(values):
+    if values == []:
+        result = None
+    else:
+        result = 0
+        for v in values:
+            if v > 0:
+                result += v
+    return result
+```
 
-### Avoid abbreviations in function and variable names.
+should instead be written:
 
-They can be ambiguous,
-and can be be hard for non-native speakers to understand.
-This rule doesn't necessarily you will have to do more typing:
-a good programming editor will [auto-complete][auto-completion] names for you.
+```python
+def positive_sum(values):
+    if values == []:
+        return None
 
-### Use short names for short-lived local variables but longer names for things with wider scope.
+    result = 0
+    for v in values:
+        if v > 0:
+            result += v
+    return result
+```
 
-Loop indices @Beni2017 can be `i` and `j` (provided the loop is only a few lines long).
-Anything that is used at a greater distance
-or whose purpose isn't immediately clear
-(such as a function) should have a longer name.
+This removes one level of nesting,
+and makes it clear what the special case is and how it's being handled.
 
-### Put everything in a file in order.
+## Where should things appear in a source code file? {#py-rse-style-order}
 
 The order of items in each file should be:
 
 -   The [shebang][shebang] line (because it has to come first to work).
--   The file's documentation string (Chapter \@ref(py-rse-package-py-document)).
+-   The file's documentation string (Section \@ref(py-rse-package-py-document)).
 -   All of the `import` statements, one per line.
--   Constant definitions.
+-   Global variable definitions
+    (especially things that would be constants in languages that support them).
 -   Function definitions.
 -   If the file can be run as a program,
     the `if __name__ == '__main__'` statement discussed in
     Section \@ref(py-rse-package-py-import).
 
-### Do not comment and uncomment sections of code to change behavior.
+That much is clear,
+but programmers disagree (strongly) on whether high-level functions should come first or last,
+i.e.,
+whether `main` should be the first function in the file or the last one.
 
-If you need to do something in some runs of the program and not do it in others,
-use an `if` statement to enable or disable that block of code.
-It's much more reliable—you're far less likely to accidentally comment out one too many lines—and
-you may find that you want to leave those conditional sections in the finished program
-for logging purposes (Section \@ref(py-rse-errors-logging)).
+We prefer the former approach,
```

Comment:
As it stands all our scripts (i.e. `countwords.py`, `collate.py` and `plotcounts.py` have `main` at the bottom of the file, so everything (the scripts and all the references to their contents them in the early chapters) will need to be updated accordingly. 

I'm thinking that could be a separate issue / PR rather than including it in this PR.

Reply:
I will adjust the comment instead.